geonode/management_commands_http/utils/job_runner.py

```diff
@@ -0,0 +1,82 @@
+#########################################################################
+#
+# Copyright (C) 2021 OSGeo
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+#########################################################################
+import io
+import json
+import sys
+
+from django.core.management import call_command
+from django.utils import timezone
+
+from geonode.management_commands_http.models import ManagementCommandJob
+
+
+class JobRunner:
+    """
+    With-statement context used to execute a ManagementCommandJob.
+    It handles status, start_time and end_time.
+    And duplicates stdout and stderr to a specified file stream.
+    """
+
+    def __init__(self, job, stream: io.StringIO, celery_result_id: str):
+        self.job = job
+        self.stream = stream
+        self.celery_result_id = celery_result_id
+
+        self.stdout = sys.stdout
+        self.stderr = sys.stderr
+
+        sys.stdout = self
+        sys.stderr = self
+
+    def write(self, data):
+        self.stream.write(data)
+        self.stdout.write(data)
+
+    def flush(self):
+        self.stream.flush()
+
+    def __enter__(self):
+        self.job.status = ManagementCommandJob.STARTED
+        if self.celery_result_id:
+            self.job.celery_result_id = self.celery_result_id
+        self.job.start_time = timezone.now()
+        self.job.save()
+        return self.job
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        sys.stdout = self.stdout
+        sys.stderr = self.stderr
+        self.job.status = ManagementCommandJob.FINISHED
+        self.job.end_time = timezone.now()
+        self.job.output_message = self.stream.getvalue()
+        self.job.save()
+
+
+def run_management_command(job_id, async_result_id: str = ""):
+    """
+    Loads the job model from database and run it using `call_command` inside a
+    context responsible to updating the status and redirecting stdout and
+    stderr.
+    """
+    job = ManagementCommandJob.objects.get(id=job_id)
```

Comment:
In this case, if the `job_id` does not exist it will raise an `Exception`. Consider adding a check, something like:
```python
job = ManagementCommandJob.objects.filter(id=job_id)
if not job.exists():
    return controlled exception or return false
 .....
other code
 .....
    with JobRunner(job.get(), output, async_result_id):
```

Reply:
Updated ✅

* Remarks:
In this case, as an approach to explicit the `Exception` case, it's better to handle the `Object.DoesNotExist` exception. 
Using `queryset.exists` and then `queryset.get`, two SQL queries will be sent to the database (and others django ORM differences). 
```
    try:
        job = ManagementCommandJob.objects.get(id=job_id)
    except ManagementCommandJob.DoesNotExist:
        raise exceptions.NotFound(
            f"ManagementCommandJob with id {job_id} was not found."
        )
```

Reply:
Yes, the exists is placed for performance reasons, it will hit the DB but as Django doc says `This tries to perform the query in the simplest and fastest way possible`
Mode details here: https://stackoverflow.com/questions/6868939/in-django-what-is-the-most-efficient-way-to-check-for-an-empty-query-set/23585057#23585057
Anyway, in this case if fine to have the Exception since we want that the exception is raised